### ISSUE TITLE
Re-Tooling Code for Page-Aligned Reading Views

### ISSUE THREAD

**Issue by ebeshero** [zme1/toscana#36]
@zme1 Here's a quick summary of the high points of our meeting today:

- [ ] You want to optimize the way your pages are aligning with your transcriptions and translations. For this we discussed:
* Setting an `@facs` on your `<pb/>` elements so that you can point to the image URL for a page in your TEI. 
* Removing or dealing with the issue of multiple `<pb/>` elements for a single page when that page ends one set of minutes and begins a new one. 
* Dealing with the overlapping hierarchy issue of pages vs. minutes. For this, we discussed "pipelining" your transformation to make the reading view. Let's try writing an XSLT Identity Transformation that flattens the XML hierarchy so that `<p>` and the like are the largest containers, and changes in minutes **as well as** changes in pages are signalled with milestone-style empty elements. This makes it easier to output a precisely page-aligned reading view, while still signalling in the HTML a shift in minutes: You'll be able to set internal links (or navigation links) to move page-by-page OR minute-by-minute. 

- [ ] Personography and Family list: Pull the information on families involved in the Lega into your personography file, by preparing a `<listOrg>` as a preceding or following sibling of your `<listPerson>`. Once you've prepared the basic listOrg, let's try writing an XSLT Identity transformation to add pointers to the persons in your listPerson who are associated with each family. Here are [the TEI specs for `<listOrg>`](http://www.tei-c.org/release/doc/tei-p5-doc/en/html/ref-listOrg.html).
* I'd just start by creating a simple `<listOrg>` with `<org>` entries and assign each an `@xml:id` like you did with the `<person>` members of `<listPerson>`. Use the `<desc>` element (as you see in the TEI example) to hold the descriptions provided by your teammates from the Google Doc. 
* Then, when you're ready, let's work out an XSLT Identity Transformation that reaches around and associates all the people connected to a family. For this we'll want to work with [the TEI element `<listRelation>`](http://www.tei-c.org/release/doc/tei-p5-doc/en/html/examples-listRelation.html), which can sit at the end of either the `<listPerson>` or the `<listOrg>`, and exists specifically to say people are identified with a family--For the encoding these are all either "passive" or "mutual" relationships b/c, well, people didn't ask to be born to X family. ;-) 

- [ ] You want to produce a network graph of the committees, with committees as edges and people as nodes.

**Comment by zme1** [comment]
@ebeshero I am sorry I didn't respond to this or work on this yesterday. I work until 5:00 today and should be able to work on this from 6:00 onward! Upon reading this writeup, especially for the hierarchical flattening, the type of output that I should try to generate is much more clearer in my mind. Thank you for this wonderful checklist!

**Comment by zme1** [comment]
@ebeshero I used Regex to add a `@facs` attribute across all my `pb` elements, but it's being flagged as invalid. I never customized any specs for my `pb`, so I'm unsure as to why it's not valid.

Here's a code snippet:

`<pb n="107" facs="../../img/107.png"/>`

Do you know why it's flagging it?

**Comment by zme1** [comment]
It looks like it's in the TEI All schema, but for some reason it is not in my customized schema, even though I didn't alter it at all. I simply declared it. I'll try to put it back in...

**Comment by ebeshero** [comment]
@zme1 I posted a question about this issue a week or so back on the TEI Stylesheets repo (about ODD processing and why the `@facs` attribute is dropped. [The answer is](https://github.com/TEIC/Stylesheets/issues/317) that you have to include the the `transcr` module to pick up the `@facs` attribute anywhere. So just add a line to your ODD to include that module, and it'll come in. (Remember to "run the wrench" to output a fresh Relax-ng XML syntax schema.)

**Comment by zme1** [comment]
@ebeshero What an easy fix. Onward!

**Comment by zme1** [comment]
@ebeshero I just pushed a reading view pipeline xml file, and the file path is `xml/meetingMinutes/volume1-reading.xml`. Would you mind looking it over and seeing if I did it as you explained? There is a `minutes` root element, and all the meetings are begun by an empty `<meeting date="date here"/>` element, and the page breaks are also all empty. I still have the transcription and translation wrapped in `div` elements, though. I'm not certain if you wanted *everything* to be entirely flattened, because I'm still foggy on the `xsl:for-each-group` that you explained to me yesterday.

**Comment by ebeshero** [comment]
@zme1 I left some comments on [your commit's XSLT file](https://github.com/zme1/toscana/commit/fe10df79d043ba0b9b852efec17ed22e106805d8#diff-adc5320e941c155f3dd9708eaf6ffdda) : You will want to flatten everything and I think you do want that to be an identity transformation, even if you're switching namespaces. 

As for the next stage in the pipeline, `<xsl:for-each-group>` helps you with flattened hierarchies--you can group elements based on XPath relationships, such as their location along the `preceding::` or `following::` axis from the element defining the group. Here's [an example XSLT using for-each-group](https://github.com/ebeshero/Pittsburgh_Frankenstein/blob/master/collateXPrep/sga_collChunking.xsl) on elements in a flattened hierarchy from the Pittsburgh Frankenstein project. In this XSLT I'm basically stopping on each `<anchor type="collate">` element and making a grouping of elements on the `following::` axis until the next `<anchor>` element, and then spitting out a new document when I do so--I call this my "coll-chunker" (like chunking coal, get it? ;-) ) because it splits up my document into new files. That's not too different from what you're working on: you *could* if you like make a new XML or HTML file around every page-beginning, or you could choose to generate any new hierarchy you need to from the flattened structure of your source input. Flattening the original hierarchies basically gives you options for negotiating the overlap you have to deal with, with pages overlapping your div structures.

**Comment by zme1** [comment]
@ebeshero According to what I gather from your comments on my XSLT file, you want to perform an identity transformation, preserving the entire document and only flattening the hierarchical elements within the bounds of each minute log? So, aside from flattening the `div` elements, nothing else will change in this pipeline. Am I interpreting your comments correctly?

**Comment by ebeshero** [comment]
@zme1 Well now--this is more about what you want to try than what I want! I noticed that you were changing to your own namespace, and I thought that was a reasonable idea since what you're creating is something that technically isn't really TEI. If you want to try that, there's a little adventure involved here with template matching on a namespace node and changing to your namespace. Now, you don't really *have* to do that, but it might be fun to try. It's up to you. 

Anyway, however you decide to this, yes, the point would simply be to flatten all the hierarchies that overlap with the `<pb/>` elements, keeping in view your goal of aligning images with their precise starting points in the output document.

**Comment by zme1** [comment]
@ebeshero I tried running an identity transformation last night, but the problem is that a flattened hierarchy in my output is invalid against my schema, which is included in the output on an identity transformation. I'm currently trying to figure out how to nix the schema lines so that I can operate outside my ODD.

**Comment by djbpitt** [comment]
@zme1 When you flatten a hierarchy en route to erecting a new one, which is what you’re doing here, the flattened one is not expected to be valid against your schema. It’s intended to be ephemeral, and as long as it’s well formed, it can serve as the input into the next transformation. You’d like the eventual final result to be valid against something, but whether that “something” is the same schema as the one that controlled your original input depends on the goal of your transformation.

**Comment by zme1** [comment]
@djbpitt Okay, that makes sense! I suppose, then, for the sake of my desire to see green again, I can just comment out the schema line and move forward.

**Comment by djbpitt** [comment]
@zme1 You can tell your transformation not to output processing instructions, which is what the schema association is syntactically, and then it won’t show up in the output. That will save you the trouble of having to comment it out; if there’s no schema association, you’ll get a green square as long as you’re well formed.

**Comment by zme1** [comment]
@ebeshero I just pushed a new reading view, which is in the same directory as the full volume of minutes. I ultimately decided not to do an identity transformation, but to declare the Toscana namespace and simply work from the ground up. I found that all the `pb` elements were nested within every type of wrapper element at some point in the volume, and it was much more succinct to just call on the elements that I want to output, rather than to tell the computer to suppress all the elements that I didn't want and to modify the ones I did.

In short, there is one `minutes` root element, with 7 `year` children (each of which has its own `@when` attribute). Inside each year are alternating, empty `div type="transcription/translation" when="yyyy-mm-dd"/>`, two for each set of minutes. Within both the translation and transcription are empty `pb` elements *only* at the points in which they appear in the manuscripts. I hope that this is somewhat close to what you were originally telling me?

**Comment by ebeshero** [comment]
@zme1 Sounds like you've got this figured out in your own way--and you know your data. What I see here is XML, so I take it this is a preliminary to constructing a new reading view for the web--and as long as you've got a system mapped out that makes sense to you and tightens up the pagination (so you're not duplicating it unnecessarily) that sounds consistent with what we discussed.

**Comment by zme1** [comment]
@ebeshero Would you be available to meet online at some point before Saturday? I do think I'm headed in the right direction thus far, but I'm venturing deeper and deeper into uncharted waters, and I think a little more face-time with this transformation could really help.

**Comment by ebeshero** [comment]
@zme1 I can try, but I'm pretty well booked tomorrow now that I'm back home in Greensburg. Sometime on Friday might work.

**Comment by zme1** [comment]
@ebeshero Alright, I plan on spending most of my next two days on this project, so just let me know!
